E2E Testing (and some testing pages in general!) Documentation Refresh (#1170)
mattermost/mattermost-developer-documentation: site/content/contribute/more-info/webapp/e2e-testing.md

---
title: "End-to-End (E2E) tests"
heading: "End-to-End (E2E) tests for the Mattermost Web app"
description: "This page describes how to run End-to-End (E2E) testing and to build tests for a section or page of the Mattermost web application."
date: "2018-03-19T12:01:23-04:00"
weight: 6
aliases:
  - contribute/webapp/e2e-testing
  - contribute/webapp/e2e
  - contribute/more-info/webapp/e2e/folder-and-file-structure
  - contribute/more-info/webapp/e2e/guide-for-writing
  - contribute/more-info/webapp/e2e/running-e2e
  - contribute/more-info/webapp/e2e/troubleshooting
---

> End-to-end tests for the Mattermost webapp in general use {{<newtabref href="https://www.cypress.io/" title="Cypress">}} and {{<newtabref href="https://playwright.dev/" title="Playwright">}}. If you're not familiar with Cypress, check out the Cypress {{<newtabref href="https://docs.cypress.io/guides/overview/why-cypress.html#In-a-nutshell" title="Developer Guide">}} and {{<newtabref href="https://docs.cypress.io/api/api/table-of-contents.html" title="API Reference">}}. Feel free to also join the {{< newtabref href="https://community.mattermost.com/core/channels/ui-test-automation" title="UI Test Automation" >}} Mattermost channel if you'd like to ask questions and collaborate with us!

Reviewer: ```suggestion
End-to-end tests for the Mattermost webapp in general use {{<newtabref href="https://www.cypress.io/" title="Cypress">}} and {{<newtabref href="https://playwright.dev/" title="Playwright">}}. If you're not familiar with Cypress, check out the Cypress {{<newtabref href="https://docs.cypress.io/guides/overview/why-cypress.html#In-a-nutshell" title="Developer Guide">}} and {{<newtabref href="https://docs.cypress.io/api/api/table-of-contents.html" title="API Reference">}}. Feel free to also join the {{< newtabref href="https://community.mattermost.com/core/channels/test-plans" title="QA Test Management" >}} Mattermost channel if you'd like to ask questions and collaborate with us!
```

Reviewer: That channel has been archived, so I suggested a different channel to link to.

* `/e2e/cypress/cypress.config.ts`: this file is for Cypress {{<newtabref href="https://docs.cypress.io/guides/references/configuration.html#Options" title="configuration">}}.
* `/e2e/cypress/package.json`: this file is for all the dependencies related to Cypress end-to-end testing.

### Writing End-to-End Tests

#### Where should a new test go?
You will need to either add the new test to an existing `spec` file, or create a new file. Sometimes, you will be informed (for example through issue descriptions) of the specific folder the test file should go in, or the actual test file being amended. As aforementioned, the `e2e/cypress/tests/integration` folder is where all of the tests live, with subdirectories that roughly divide the tests by functional areas. Cypress is configured to look for and run tests that match the pattern of `*_spec.ts`, so a good new test file name for an issue like {{<newtabref href="https://github.com/mattermost/mattermost-server/issues/18184" title=`Write Webapp E2E with Cypress: "MM-T642 Attachment does not collapse" #18184`>}} would be `attachment_does_not_collapse_spec.ts`, to ensure that it gets picked up.

> > *Note*: There may be some JavaScript `spec` files, but new tests should be written in TypeScript - if you are adding a test to an existing `spec` file, convert that file to TypeScript if necessary.

Reviewer: ```suggestion
> *Note*: There may be some JavaScript `spec` files, but new tests should be written in TypeScript. If you are adding a test to an existing `spec` file, convert that file to TypeScript if necessary.
```

  ```javascript
  describe('Integrations/Incoming Webhook', () => {
    it('MM-T642 Attachment does not collapse', () => {
      // Put test steps and assertions here
    });
  });
  ```
For those writing E2E from Help Wanted tickets with `Area/E2E Tests` label, the `Test Key` is available in the Github issue itself. The `Test Key` is used for mapping test cases per Release Testing specification. It will be used to measure coverage between manual and automated tests. In case the `Test Key` is not available, feel free to prompt the QA team who will either search for an existing Zephyr entry or if it's a new one, it will be created for you.

#### Using Cypress Hooks

> Before writing the main body of the test in the `it` block, it can help to write some setup code for test isolation using {{<newtabref href="https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Hooks" title="hooks">}}. In a `before()` hook, you can run tests in isolation using the custom command `cy.apiInitSetup()`. This command creates a new team, channel, and user which can only be used by the spec file itself. Make use of the `cy.apiInitSetup()` function as much as possible, as it is recommended to log in as a new user and visit the generate team and/or channel. Avoid the use of `sysadmin` user or default `ad-1` team if possible.

Reviewer: ```suggestion
Before writing the main body of the test in the `it` block, it can help to write some setup code for test isolation using {{<newtabref href="https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Hooks" title="hooks">}}. In a `before()` hook, you can run tests in isolation using the custom command `cy.apiInitSetup()`. This command creates a new team, channel, and user which can only be used by the spec file itself. Make use of the `cy.apiInitSetup()` function as much as possible, as it is recommended to log in as a new user and visit the generated team and/or channel. Avoid the use of `sysadmin` user or default `ad-1` team if possible.
```

Review the [Test Guidelines]({{<ref "/contribute/more-info/getting-started/test-guideline">}}) for details on how to submit your PR

### Troubleshooting
#### Test(s) failing due to a known issue
If test(s) are failing due to another known issue, follow these steps to amend your test:
1. Append the Jira issue key in the test title, following the format of ` -- KNOWN ISSUE: [Jira_key]`. For example:
    ```javascript
    describe('Upload Files', () => {
      it('MM-T2261 Upload SVG and post -- KNOWN ISSUE: MM-38982', () => {
        // Test steps and assertion here
      }
    }
    ```
2. Move the test case into a separate `spec` file following the format of `<existing_spec_file_name_[1-9].js>`. For example:
     `accessibility_account_settings_spec_1.js` and demote the spec file (i.e. remove `// Stage: @prod` from the spec file)
3. If all the test cases are failing in a spec file, update each title as mentioned above and demote the spec file.
4. Link the failed test case(s) to the Jira issue (the known issue). In the Jira bug, select the **Zephyr Scale** tab. Select the **add an existing one** link, then select test case(s), and finally select **Add**.

> 5. Conversely, remove the Jira issue key if the issue has been resolved and the test is passing.

Reviewer 2: Thanks @furqanmlk! Please have it in a separate follow up PR.
